Support sheet table function for google sheets connector

grantatspothero opened the pull request:
<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->
<!-- Provide an overview of the PR for maintainers and reviewers. -->
## Description
Supports `sheet` table function for the google sheets connector. 

Allows querying google sheets directly by id instead of requiring a metadata sheet mapping. 

Also makes the `gsheets.metadata-sheet-id` config optional since the sheet table function makes it no longer strictly required.  


<!-- Provide details that would help an engineer who is unfamiliar with this part of the code. -->
## Additional context and related issues
N/A


<!-- Mark the appropriate option with an (x). Propose a release note if you can. -->
## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
# Google Sheets
* Add `sheet` table function to the connector. ({issue}`12502`)
```

grantatspothero commented on plugin/trino-google-sheets/src/main/java/io/trino/plugin/google/sheets/SheetsSplitManager.java:
```diff
@@ -52,12 +53,19 @@ public ConnectorSplitSource getSplits(
+                        return new TableNotFoundException(tableHandle.asPlainTable().toSchemaTableName());
+                    }
+                    else {
+                        return new TableNotFoundException(new SchemaTableName("_generated", tableHandle.asSheetTable().getSheetExpression()));
```
Was not sure what to do here since sheet tables do not have a schema or table name associated with them.

kasiafi replied:
Maybe add another subclass of `NotFoundException`... Reporting a synthesized name as the schema name, and a sheet id in place of a table name might be misleading.

martint commented on docs/src/main/sphinx/connector/googlesheets.rst:
```diff
@@ -133,3 +133,28 @@ SQL support
+      *
+    FROM
+      TABLE(
+        sheets.system.query(
```
This is not a "query", so we need a better name for the functions. Some possibilities include "read" or "sheet".

grantatspothero replied:
Sure will switch to "sheet"

grantatspothero replied:
@martint "id" is not the best name since the parameter is actually an expression.

The expression consists of `${sheet_id}#${sheet_range}`. the default sheet range if one is not provided is `$1:$10000`, but the sheet range can also refer to [named ranges](https://support.google.com/docs/answer/63175?hl=en&co=GENIE.Platform%3DDesktop)

"sheet" is better, but `sheets.system.sheet(sheet => 'blah#blah')` is a mouthful 🤷

martint replied:
How about `range` then, since that's what that expression represents?

kasiafi commented on plugin/trino-google-sheets/src/main/java/io/trino/plugin/google/sheets/SheetsTableHandle.java:
```diff
@@ -14,50 +14,62 @@
         implements ConnectorTableHandle
 {
-    private final SchemaTableName schemaTableName;
+    private final SheetsRelationHandle relationHandle;
```
Instead of making `SheetsTableHandle` a "two-in-one", we could add another implementation of `ConnectorTableHandle` for the purpose of the Table Function. The code would be clearer that way, and we would avoid nesting / optionality which makes it harder to reason about.

grantatspothero replied:
Fixed to remove nesting, take another look.

kasiafi commented on plugin/trino-google-sheets/src/main/java/io/trino/plugin/google/sheets/SheetsMetadata.java:
```diff
@@ -145,4 +159,22 @@ public ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTable
+    @Override
+    public Optional<TableFunctionApplicationResult<ConnectorTableHandle>> applyTableFunction(ConnectorSession session, ConnectorTableFunctionHandle handle)
+    {
+        if (!(handle instanceof Query.QueryFunctionHandle)) {
+            return Optional.empty();
+        }
+
+        ConnectorTableHandle tableHandle = ((Query.QueryFunctionHandle) handle).getTableHandle();
+        ConnectorTableSchema tableSchema = getTableSchema(session, tableHandle);
+        Map<String, ColumnHandle> columnHandlesByName = getColumnHandles(session, tableHandle);
+        List<ColumnHandle> columnHandles = tableSchema.getColumns().stream()
+                .map(ColumnSchema::getName)
+                .map(columnHandlesByName::get)
+                .collect(toImmutableList());
+
+        return Optional.of(new TableFunctionApplicationResult<>(tableHandle, columnHandles));
```
Please revert the if condition.

grantatspothero replied:
I copied this pattern from `applyTableFunction` in other connectors, see bigquery and base-jdbc:
https://github.com/trinodb/trino/blob/master/plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java#L701-L704
https://github.com/trinodb/trino/blob/master/plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java#L606-L608

kasiafi replied:
I think we'll eventually revert as we add more Table Functions. You can leave it for now.

kasiafi commented on plugin/trino-google-sheets/src/main/java/io/trino/plugin/google/sheets/SheetsSplit.java:
```diff
@@ -33,35 +33,17 @@
 {
     private static final int INSTANCE_SIZE = toIntExact(ClassLayout.parseClass(SheetsSplit.class).instanceSize());
 
-    private final String schemaName;
```
>Alternatively, we could keep a name property and fill in the sheetId instead.

The semantics of this field should not depend on the context.

You can add a dedicated implementation of `ConnectorSplit` which would have the `id` field. Or add another field here, and make them all optional. 
i'm leaning towards the latter, because the semantics of the Split is the same in both cases, and they only differ in the available metadata.

grantatspothero replied:
fixed.

kasiafi commented on plugin/trino-google-sheets/src/main/java/io/trino/plugin/google/sheets/ptf/Query.java:
```diff
@@ -0,0 +1,127 @@
+                    NAME,
+                    ImmutableList.of(
+                            ScalarArgumentSpecification.builder()
+                                    .name("SHEET")
```
When you rename the function to `sheet`, it could be good to rename the argument to `id`.

kasiafi commented on plugin/trino-google-sheets/src/test/java/io/trino/plugin/google/sheets/SheetsQueryRunner.java:
```diff
@@ -35,22 +36,23 @@ private SheetsQueryRunner() {}
-            connectorProperties.putIfAbsent("gsheets.credentials-path", getTestCredentialsPath());
-            connectorProperties.putIfAbsent("gsheets.metadata-sheet-id", TEST_METADATA_SHEET_ID);
-            connectorProperties.putIfAbsent("gsheets.max-data-cache-size", "1000");
-            connectorProperties.putIfAbsent("gsheets.data-cache-ttl", "5m");
+            final Map<String, String> properties = new HashMap<>(ImmutableMap.copyOf(connectorProperties));
+            properties.putIfAbsent("gsheets.credentials-path", getTestCredentialsPath());
+            metadataSheetId.ifPresent(sheetId -> properties.putIfAbsent("gsheets.metadata-sheet-id", sheetId));
+            properties.putIfAbsent("gsheets.max-data-cache-size", "1000");
+            properties.putIfAbsent("gsheets.data-cache-ttl", "5m");
 
             queryRunner.installPlugin(new SheetsPlugin());
-            queryRunner.createCatalog(GOOGLE_SHEETS, GOOGLE_SHEETS, connectorProperties);
+            queryRunner.createCatalog(GOOGLE_SHEETS, GOOGLE_SHEETS, properties);
```
Please revert the rename, or extract it as another commit.

grantatspothero replied:
reverted the rename.

grantatspothero commented on plugin/trino-google-sheets/src/main/java/io/trino/plugin/google/sheets/SheetsClient.java:
```diff
@@ -133,8 +150,8 @@ public Optional<SheetsTable> getTable(String tableName)
-            List<List<String>> dataValues = values.subList(1, values.size()); // removing header info
-            return Optional.of(new SheetsTable(tableName, columns.build(), dataValues));
+            List<List<String>> dataValues = stringValues.subList(1, stringValues.size()); // removing header info
+            return Optional.of(new SheetsTable(tableName.orElse("_generated"), columns.build(), dataValues));
```
Note here in `SheetsTable` the tableName is required, but is completely unused.

Is it worth removing the tableName from `SheetsTable`?

ebyhr replied:
I agree with removing the variable.

grantatspothero replied:
removed.

grantatspothero commented:
Ready for another review.

kasiafi commented on plugin/trino-google-sheets/src/main/java/io/trino/plugin/google/sheets/SheetsConnectorTableHandle.java:
```diff
@@ -16,27 +16,55 @@
         return this instanceof SheetsNamedTableHandle;
     }
 
+    default boolean isSheetTable()
```
I don't think that the methods `isSheetTable()` etc. add to readability. I would just use explicit class checks and casts instead. I'm not sure we even need the enclosing interface.

grantatspothero replied:
When [pattern matching ](https://openjdk.org/jeps/427)comes out of preview, the sealed interface permits exhaustive pattern matching on the implementing interfaces. 
 
I think the helper methods adds readability over lots of instanceof and casts everywhere, but readability is more subjective.

Can revert if you feel strongly about it.

kasiafi commented on plugin/trino-google-sheets/src/test/java/io/trino/plugin/google/sheets/TestGoogleSheets.java:
```diff
@@ -78,4 +80,39 @@ public void testTableWithRepeatedAndMissingColumnNames()
+    public void testSheetQueryWithSheetTab()
+    {
+        assertQuery(
+                "SELECT * FROM TABLE(gsheets.system.sheet(id => '%s'))".formatted(DATA_SHEET_ID + "#number_text"),
```
is `"#number_text"` the range? Could you add a test where the range affects the results?

grantatspothero replied:
Sure, unfortunately this requires permissions to modify the existing testing google sheet and I do not have permissions to do so.

https://docs.google.com/spreadsheets/d/1S625j2oTptRepg6Yci68fCYE1269tdoSjljNOmTgQ3U/edit#gid=0

grantatspothero replied:
I added more tests using the same sheet but with a different range filter. This tests the functionality without requiring modifying the sheet.

grantatspothero commented on docs/src/main/sphinx/connector/googlesheets.rst:
```diff
@@ -133,3 +133,41 @@ SQL support
+      );
+
+
+A sheet range or named range can be provided after the Google Sheet id to only access a subset of data in a sheet.
+The default sheet range is ``$1:$10000`` if one is not provided::
```
Agree with you, but this is how the connector already works. In the current `metadata-sheet-id` you reference other sheets by `${sheet_id}#${sheet_range}`

For consistency purposes I thought it made sense to keep the pattern the same. 

What do you think?

martint replied:
@kasiafi’s suggestion makes a lot of sense and it’s a much more natural way to make these arguments. This function is independent from how the table mappings work, so there’s no need to conform to that pattern.

grantatspothero replied:
Sure will do.

ebyhr commented on plugin/trino-google-sheets/src/test/java/io/trino/plugin/google/sheets/TestGoogleSheets.java:
```diff
@@ -78,4 +83,39 @@ public void testTableWithRepeatedAndMissingColumnNames()
+    // test polymorphic table function
+
+    @Test
+    public void testSheetQuerySimple()
```
Could you add another test method having outer WHERE clause?

grantatspothero replied:
Added another test with a where clause to validate column naming.

ebyhr commented on plugin/trino-google-sheets/src/main/java/io/trino/plugin/google/sheets/SheetsConnectorTableHandle.java:
```diff
@@ -0,0 +1,44 @@
+        extends ConnectorTableHandle
+        permits SheetsNamedTableHandle, SheetsSheetTableHandle
+{
+    default NotFoundException tableNotFound()
```
Putting this method in the interface looks weird to me.

grantatspothero replied:
Where would you like the method to exist instead? Making it a static method in a utility class seems worse.

ebyhr replied:
Probably, we can change `Optional<SheetsTable> SheetsClient.getTable(SheetsConnectorTableHandle tableHandle)`  to non-optional and move this method to around there. If we still want to use an interface, moving the implementation to `SheetsNamedTableHandle` to `SheetsSheetTableHandle` looks better to me.

grantatspothero replied:
Moved the implementation to the implementing classes. Removing the optional is more difficult since `SheetsMetadata#getTableHandle` returns null when the optional is empty while other methods throw.

grantatspothero commented on plugin/trino-google-sheets/src/test/java/io/trino/plugin/google/sheets/SheetsQueryRunner.java:
```diff
@@ -35,7 +36,8 @@ private SheetsQueryRunner() {}
             Map<String, String> extraProperties,
-            Map<String, String> connectorProperties)
+            Map<String, String> connectorProperties,
+            Optional<String> metadataSheetId)
```
Sure, can revert not strictly necessary.

grantatspothero commented on plugin/trino-google-sheets/src/main/java/io/trino/plugin/google/sheets/SheetsClient.java:
```diff
@@ -170,6 +187,17 @@ public List<List<Object>> readAllValues(String tableName)
         }
     }
 
+    public List<List<Object>> readAllValuesFromSheet(String sheetExpression)
```
done.

grantatspothero commented on plugin/trino-google-sheets/src/main/java/io/trino/plugin/google/sheets/SheetsConnectorFactory.java:
```diff
@@ -44,7 +49,11 @@ public Connector create(String catalogName, Map<String, String> config, Connecto
                 new TypeDeserializerModule(context.getTypeManager()),
-                new SheetsModule());
+                new SheetsModule(),
+                (Binder binder) -> newSetBinder(binder, ConnectorTableFunction.class)
+                        .addBinding()
+                        .toProvider(Sheet.class)
+                        .in(Scopes.SINGLETON));
```
moved.

grantatspothero commented on plugin/trino-google-sheets/src/main/java/io/trino/plugin/google/sheets/SheetsMetadata.java:
```diff
@@ -127,6 +131,15 @@ private Optional<ConnectorTableMetadata> getTableMetadata(SchemaTableName tableN
         return Optional.empty();
     }
 
+    private Optional<ConnectorTableMetadata> getTableMetadata(SheetsConnectorTableHandle tableHandle)
+    {
+        Optional<SheetsTable> table = sheetsClient.getTable(tableHandle);
+        if (table.isPresent()) {
+            return Optional.of(new ConnectorTableMetadata(getSchemaTableName(tableHandle), table.get().getColumnsMetadata()));
+        }
+        return Optional.empty();
+    }
```
inlined.

grantatspothero commented on plugin/trino-google-sheets/src/main/java/io/trino/plugin/google/sheets/SheetsMetadata.java:
```diff
@@ -145,4 +158,34 @@ public ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTable
+
+    private static SchemaTableName getSchemaTableName(SheetsConnectorTableHandle handle)
+    {
+        // TODO (https://github.com/trinodb/trino/issues/6694) SchemaTableName should not be required for synthetic ConnectorTableHandle
```
`isSheetTable` was removed. I think this is the proper place for the comment.

ebyhr replied:
The comment relates to `SheetsSheetTableHandle`, not  `SheetsNamedTableHandle`. I would move to the below place. 
```java
        if (handle instanceof SheetsSheetTableHandle) {
            // move to here
            return new SchemaTableName("_generated", "_generated");
        }
```

grantatspothero replied:
Ahh gotcha. will fix.

grantatspothero commented on plugin/trino-google-sheets/src/main/java/io/trino/plugin/google/sheets/ptf/Sheet.java:
```diff
@@ -0,0 +1,128 @@
+                    .map(entry -> new Descriptor.Field(
+                            entry.getKey(),
+                            Optional.of(((SheetsColumnHandle) (entry.getValue())).getColumnType())))
+                    .collect(Collectors.toList());
```
fixed.

grantatspothero commented on docs/src/main/sphinx/connector/googlesheets.rst:
```diff
@@ -133,3 +133,41 @@ SQL support
 The connector provides :ref:`globally available <sql-globally-available>` and
 :ref:`read operation <sql-read-operations>` statements to access data and
 metadata in Google Sheets.
```
done.

grantatspothero commented on plugin/trino-google-sheets/src/test/java/io/trino/plugin/google/sheets/TestGoogleSheets.java:
```diff
@@ -78,4 +83,39 @@ public void testTableWithRepeatedAndMissingColumnNames()
+    public void testSheetQueryWithInvalidSheetId()
+    {
+        assertThatThrownBy(() -> query("SELECT * FROM TABLE(gsheets.system.sheet(id => 'DOESNOTEXIST'))"))
+                .hasMessageContaining("Failed reading data from sheet: DOESNOTEXIST");
```
This was the existing behavior in the connector, see `readAllValuesFromSheetExpression`:
```
throw new TrinoException(SHEETS_UNKNOWN_TABLE_ERROR, "Failed reading data from sheet: " + sheetExpression, e);
```

I wasn't sure of the consequences of changing the exception thrown.

grantatspothero replied:
Added a comment.

ebyhr commented on plugin/trino-google-sheets/src/test/java/io/trino/plugin/google/sheets/TestGoogleSheets.java:
```diff
@@ -78,4 +85,48 @@ public void testTableWithRepeatedAndMissingColumnNames()
+    }
+
+    @Test
+    public void testSheetQueryWithRange()
```
Range allows narrowing columns, let's add the test case. e.g. `range => 'number_text!B3:B5'`

grantatspothero replied:
added.

grantatspothero commented on plugin/trino-google-sheets/src/test/java/io/trino/plugin/google/sheets/TestGoogleSheets.java:
```diff
@@ -78,4 +85,48 @@ public void testTableWithRepeatedAndMissingColumnNames()
+                        "('2', 'two')," +
+                        "('3', 'three')," +
+                        "('4', 'four')," +
+                        "('5', 'five')");
```
> ```sql
> $1:$2
> ```

This is intended will document the behavior. The first row of the range is treated as the header.

grantatspothero replied:
> Also, the column definition differs based on the provided range. I htiWe should get the column names form the 1st row regardless of range, or document this behavior.
> 
> ```sql
> trino> SELECT
>     ->       *
>     ->     FROM
>     ->       TABLE(
>     ->         gsheets.system.sheet(
>     ->           id => '1S625j2oTptRepg6Yci68fCYE1269tdoSjljNOmTgQ3U', range => '$2:$2'
>     ->         )
>     ->       );
>  1 | one   <- different from "number | text" 
> ---+-----
> (0 rows)
> ```

This is intended will document the behavior. The first row of the range is treated as the header.

grantatspothero replied:
Added validation to the `id` argument and documented the behavior of using the first row of the range as the header row.

grantatspothero commented on plugin/trino-google-sheets/src/main/java/io/trino/plugin/google/sheets/SheetsConnector.java:
```diff
@@ -34,18 +37,21 @@
         this.metadata = requireNonNull(metadata, "metadata is null");
         this.splitManager = requireNonNull(splitManager, "splitManager is null");
         this.recordSetProvider = requireNonNull(recordSetProvider, "recordSetProvider is null");
+        this.connectorTableFunctions = requireNonNull(connectorTableFunctions, "connectorTableFunctions is null");
```
fixed.

grantatspothero commented on plugin/trino-google-sheets/src/main/java/io/trino/plugin/google/sheets/SheetsSheetTableHandle.java:
```diff
@@ -0,0 +1,81 @@
+    private final String sheetRange;
+
+    @JsonCreator
+    public SheetsSheetTableHandle(@JsonProperty("sheetExpression") String sheetId, String sheetRange)
```
fixed.

grantatspothero commented on docs/src/main/sphinx/connector/googlesheets.rst:
```diff
@@ -133,3 +133,42 @@ SQL support
+      *
+    FROM
+      TABLE(
+        sheets.system.sheet(
```
fixed.

grantatspothero commented:
@ebyhr @kasiafi just a nudge for another review.

grantatspothero commented on plugin/trino-google-sheets/src/main/java/io/trino/plugin/google/sheets/SheetsNamedTableHandle.java:
```diff
@@ -15,20 +15,19 @@
 
-public final class SheetsTableHandle
-        implements ConnectorTableHandle
+public final class SheetsNamedTableHandle
```
Good catch, removed. Should there be a lint rule for unused methods in CI?

grantatspothero commented on plugin/trino-google-sheets/src/main/java/io/trino/plugin/google/sheets/SheetNotFoundException.java:
```diff
@@ -0,0 +1,27 @@
+{
+    public SheetNotFoundException(String sheetExpression)
+    {
+        super(format("Sheet '%s' not found", sheetExpression));
```
This exception only gets thrown if a google sheet is removed or modified in a way that causes failure during processing of a query. 

Otherwise, this exception will be thrown when attempting to originally read the sheet range:
`TrinoException(SHEETS_UNKNOWN_TABLE_ERROR, "Sheet expression not found for table " + tableName)`

That is difficult to integration test with an actual sheet, but I can at least exercise this code path with a unit test.

ebyhr replied:
No need to add unit tests in my opinion.

anusudarsan commented on plugin/trino-google-sheets/src/main/java/io/trino/plugin/google/sheets/ptf/Sheet.java:
```diff
@@ -0,0 +1,152 @@
+        // Ranges should be provided through the range argument
+        // https://developers.google.com/sheets/api/guides/concepts
+        if (sheetId.contains(RANGE_SEPARATOR)) {
+            throw new TrinoException(INVALID_FUNCTION_ARGUMENT, "Google sheet ID %s cannot contain '#'. Provide a range through the 'range' argument.".formatted(sheetId));
```
`"Google sheet ID %s cannot contain '#'.
`
is this already documented?

grantatspothero replied:
See the link: https://developers.google.com/sheets/api/guides/concepts Google sheet ids by definition cannot contain '#'. 

Trino uses '#' as a separator character in the metadata sheet mapping so people can query tabs and ranges within a sheet. For example:) `sheet_id#tab_name!D12:E13`

This error message is because the sheet function has separated named parameters for the id and range respectively while the metadata sheet uses a single field with a '#' separator between the id and range. 

If someone puts a '#' in the id field, it is an error.

ebyhr commented:
Could you confirm and fix the commit boundaries? `check-commit` failed due to a missing symbol.

grantatspothero commented:
> Could you confirm and fix the commit boundaries? `check-commit` failed due to a missing symbol.

Removed `toSchemaTableName` in one commit too early, fixed.

anusudarsan commented:
@kasiafi or @martint can you please do a final review/approve this?

grantatspothero commented:
@wendigo mind giving a final set of eyes? @ebyhr already gave a detailed review.

findepi commented on plugin/trino-google-sheets/src/main/java/io/trino/plugin/google/sheets/SheetsSheetTableHandle.java:
```diff
@@ -0,0 +1,90 @@
+    @Override
+    public int hashCode()
+    {
+        return Objects.hash(sheetId);
```
sheetRange?

findepi commented on plugin/trino-google-sheets/src/main/java/io/trino/plugin/google/sheets/SheetsSheetTableHandle.java:
```diff
@@ -0,0 +1,90 @@
+    @Override
+    public String toString()
+    {
+        return format("Sheet[%s]", sheetId);
```
sheetRange?

grantatspothero replied:
Good catch, originally the sheetRange was part of the sheetId but when I did the refactor to separate them I forgot to update here. Fixed.

findepi commented on plugin/trino-google-sheets/src/test/java/io/trino/plugin/google/sheets/TestGoogleSheets.java:
```diff
@@ -78,4 +80,114 @@ public void testTableWithRepeatedAndMissingColumnNames()
                 " ('column_1','varchar','',''), ('column_2','varchar','',''), ('c','varchar','',''))");
     }
+
+    // test polymorphic table function
```
we don't use block comments like this. they always flow away from a block of code they are decorating (or the block of code gains new members mixed in)

grantatspothero replied:
Removed.

findepi commented on plugin/trino-google-sheets/src/main/java/io/trino/plugin/google/sheets/SheetsClient.java:
```diff
@@ -161,8 +161,11 @@ public Optional<SheetsTable> getTableFromValues(List<List<Object>> values)
     {
         ImmutableSet.Builder<String> tables = ImmutableSet.builder();
+        if (metadataSheetId.isEmpty()) {
+            return tables.build();
```
`return ImmutableList.of();` and move the builder allocation below this check

findepi replied:
nit... but bump :)

findepi commented on plugin/trino-google-sheets/src/main/java/io/trino/plugin/google/sheets/SheetsClient.java:
```diff
@@ -219,7 +222,10 @@ private Optional<String> getSheetExpressionForTable(String tableName)
         ImmutableMap.Builder<String, Optional<String>> tableSheetMap = ImmutableMap.builder();
-        List<List<Object>> data = readAllValuesFromSheetExpression(metadataSheetId);
+        if (metadataSheetId.isEmpty()) {
+            return tableSheetMap.buildOrThrow();
```
as above, `return ImmutableMap.of()`

findepi replied:
.. and here

findepi commented on plugin/trino-google-sheets/src/test/java/io/trino/plugin/google/sheets/TestGoogleSheetsWithoutMetadataSheetId.java:
```diff
@@ -0,0 +1,46 @@
+import static io.trino.plugin.google.sheets.SheetsQueryRunner.createSheetsQueryRunner;
+import static io.trino.plugin.google.sheets.TestSheetsPlugin.DATA_SHEET_ID;
+
+public class TestGoogleSheetsWithoutMetadataSheetId
```
How long does it take to run this test class?

grantatspothero replied:
Very fast, ~15 seconds on my local machine.

findepi replied:
what about adding a test case in TestSheetsPlugin (instead?)

grantatspothero replied:
I do not think that helps, `TestSheetsPlugin` runs in 1-2 seconds currently. 

Moving the test case in there will just make `TestSheetsPlugin` take ~15 seconds instead of ~1-2 seconds.

findepi replied:
@grantatspothero i understand the approach currently taken is best from test coverage perspective.
@nineinchnick and I were recently debating our inherently limited ability to have test coverage, and how much we can test in total on the CI. Looks like there are and will always be limits to that. This made me more cautious about time being added to tests with respect to additional confidence gained when doing so. Ultimately, we need to accept the fact we cannot test everything, which is especially sad to me.

(no change requested here, but I'd still consider not having this test class)

mosabua commented on docs/src/main/sphinx/connector/googlesheets.rst:
```diff
@@ -133,3 +137,42 @@ SQL support
+
+For example::
+
+    SELECT
+      *
+    FROM
+      TABLE(
+        example.system.sheet(
```
Following the Mozilla SQL style ideally .. in this case I would confirmhow we format in all other docs .. but I think this is right

grantatspothero replied:
I copied this formatting from the other table function docs, example: https://github.com/trinodb/trino/blob/master/docs/src/main/sphinx/connector/mongodb.rst#table-functions

findepi replied:
is this formatting compatible with how Trino SQL formatter would format this?

grantatspothero replied:
Did this in a debugger:
```
var statement = new SqlParser().createStatement("SELECT * FROM TABLE(example.system.sheet(id => 'googleSheetIdHere', range => 'TabName!A1:B4'))", new ParsingOptions());

SqlFormatter.formatSql(statement)
```

The formatter decides to write the query like this:
```
SELECT *
FROM
  TABLE(example.system.sheet(
      id => 'googleSheetIdHere',
      range => 'TabName!A1:B4'))
```

Will change.

grantatspothero commented on plugin/trino-google-sheets/src/main/java/io/trino/plugin/google/sheets/SheetsConnectorTableHandle.java:
```diff
@@ -16,15 +16,18 @@
 @JsonSubTypes({
-        @JsonSubTypes.Type(value = SheetsNamedTableHandle.class, name = "named")
+        @JsonSubTypes.Type(value = SheetsNamedTableHandle.class, name = "named"),
+        @JsonSubTypes.Type(value = SheetsSheetTableHandle.class, name = "sheet")
```
Will change.

The style checker isn't forcing usages of trailing `,` is that something that needs to change?

findepi replied:
you mean that `@JsonSubTypes` is getting removed? right

grantatspothero commented on docs/src/main/sphinx/connector/googlesheets.rst:
```diff
@@ -133,3 +137,42 @@ SQL support
+    FROM
+      TABLE(
+        example.system.sheet(
+          id => 'googleSheetIdHere'
```
See `TestSheetsPlugin` for an example sheet id. For example: `1S625j2oTptRepg6Yci68fCYE1269tdoSjljNOmTgQ3U`

grantatspothero commented on plugin/trino-google-sheets/src/test/java/io/trino/plugin/google/sheets/TestSheetsConnectorTableHandle.java:
```diff
@@ -0,0 +1,35 @@
+
+public class TestSheetsConnectorTableHandle
+{
+    private final JsonCodec<SheetsConnectorTableHandle> codec = JsonCodec.jsonCodec(SheetsConnectorTableHandle.class);
```
I just confirmed, you are correct this bit is not necessary. 

Will remove and just test that the specific table handles can be serialized/deserialized.

grantatspothero commented on plugin/trino-google-sheets/src/test/java/io/trino/plugin/google/sheets/TestSheetsConnectorTableHandle.java:
```diff
@@ -0,0 +1,35 @@
+        String json = codec.toJson(expected);
+        SheetsConnectorTableHandle actual = codec.fromJson(json);
+
+        assertEquals((SheetsNamedTableHandle) actual, (SheetsNamedTableHandle) expected);
```
Removed after the above refactor.

grantatspothero commented on plugin/trino-google-sheets/src/main/java/io/trino/plugin/google/sheets/SheetsClient.java:
```diff
@@ -162,14 +181,25 @@ public List<List<Object>> readAllValues(String tableName)
             String sheetExpression = tableSheetMappingCache.getUnchecked(tableName)
                     .orElseThrow(() -> new TrinoException(SHEETS_UNKNOWN_TABLE_ERROR, "Sheet expression not found for table " + tableName));
-            return sheetDataCache.getUnchecked(sheetExpression);
+            return readAllValuesFromSheet(sheetExpression);
         }
         catch (UncheckedExecutionException e) {
             throwIfInstanceOf(e.getCause(), TrinoException.class);
             throw new TrinoException(SHEETS_TABLE_LOAD_ERROR, "Error loading data for table: " + tableName, e);
```
`tableSheetMappingCache` on the previous line can still throw `UncheckedExecutionException`, I think this is still needed.

findepi replied:
you're right. it would be appropriate to pull readAllValuesFromSheet after try-catch block then

grantatspothero replied:
Doing that makes the code more complex since `sheetExpression` would have to be declared outside of the try block. 

IMO I think keeping it as is makes the most sense.

findepi replied:
> Doing that makes the code more complex since `sheetExpression` would have to be declared outside of the try block.

correct

grantatspothero commented on plugin/trino-google-sheets/src/main/java/io/trino/plugin/google/sheets/SheetsClient.java:
```diff
@@ -245,6 +275,7 @@ private List<List<Object>> readAllValuesFromSheetExpression(String sheetExpressi
             return sheetsService.spreadsheets().values().get(sheetId, defaultRange).execute().getValues();
         }
         catch (IOException e) {
+            // TODO: improve error to a {Table|Sheet}NotFoundException
```
See comment from @ebyhr here: https://github.com/trinodb/trino/pull/15829#discussion_r1092600469

I did not want to change the error code returned from `SHEETS_UNKNOWN_TABLE_ERROR` to `NOT_FOUND` since this would be backwards incompatible. 

Updated the comment to make the reasoning more clear.

findepi replied:
thanks, now i can read the comment at least :) 

however, i don't understand why we would want to throw "not found" within IOException catch block
to me, the current SHEETS_UNKNOWN_TABLE_ERROR looks fine (the IOException is just some error)

if we want to detect that IOException conveys "specific sheet doesn't exist", we could do that (but we wouldnt' say TableNotFound even then, because the requested sheet may not be a table)


if it was up to me to decide, and if no new information is provided to me, i would just delete this TODO comment and worry about that later, and only if we learn later there is something to worry about

grantatspothero commented on plugin/trino-google-sheets/src/main/java/io/trino/plugin/google/sheets/SheetsConnectorTableHandle.java:
```diff
@@ -16,15 +16,18 @@
         use = JsonTypeInfo.Id.NAME,
         property = "@type")
 @JsonSubTypes({
-        @JsonSubTypes.Type(value = SheetsNamedTableHandle.class, name = "named")
```
Not necessary after the refactor you proposed.

grantatspothero commented on plugin/trino-google-sheets/src/main/java/io/trino/plugin/google/sheets/SheetsConnectorTableHandle.java:
```diff
@@ -16,15 +16,18 @@
-        permits SheetsNamedTableHandle
+        permits SheetsNamedTableHandle, SheetsSheetTableHandle
 {
+    NotFoundException tableNotFound();
```
That is fair, moved to a static default method on the interface instead.

grantatspothero commented on plugin/trino-google-sheets/src/main/java/io/trino/plugin/google/sheets/SheetsMetadata.java:
```diff
@@ -79,17 +83,19 @@ public SheetsNamedTableHandle getTableHandle(ConnectorSession session, SchemaTab
-        return getTableMetadata(tableHandle.toSchemaTableName())
-                .orElseThrow(() -> new TrinoException(SHEETS_UNKNOWN_TABLE_ERROR, "Metadata not found for table " + tableHandle.getTableName()));
+        SheetsConnectorTableHandle tableHandle = (SheetsConnectorTableHandle) table;
+        Optional<SheetsTable> sheetsTable = sheetsClient.getTable(tableHandle);
+        return sheetsTable
+                .map(unwrappedTable -> new ConnectorTableMetadata(getSchemaTableName(tableHandle), unwrappedTable.getColumnsMetadata()))
+                .orElseThrow(() -> new TrinoException(SHEETS_UNKNOWN_TABLE_ERROR, "Metadata not found for table " + tableHandle.tableNotFound()));
```
Agreed is simpler, fixed.

findepi commented on plugin/trino-google-sheets/src/main/java/io/trino/plugin/google/sheets/SheetsClient.java:
```diff
@@ -235,16 +271,22 @@ private List<List<Object>> readAllValuesFromSheetExpression(String sheetExpressi
         }
         catch (IOException e) {
+            // TODO: improve error to a {Table|Sheet}NotFoundException
+            // after confirming the backwards incompatible error code change from 2->5 is OK
```
what 2 and 5 are referring to?

grantatspothero replied:
The error codes for: SHEETS_UNKNOWN_TABLE_ERROR and NOT_FOUND respectively. changed the comment to use the names instead of the codes to be clearer.

grantatspothero commented on docs/src/main/sphinx/connector/googlesheets.rst:
```diff
@@ -133,3 +137,42 @@ SQL support
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The ``sheet`` function allows you to query a Google Sheet directly without
+specifying it as a named table in the metadata sheet.
```
fixed.

findepi reviewed (approved):
LGTM

thanks @grantatspothero for quick response loop

findepi commented:
(rebased to resolve conflict)

findepi commented:
/test-with-secrets sha=d3ed6c25f1b3add9cdd97d9590780e24ecc3a38d

findepi commented:
CI green. Waiting for secrets run to finish.

grantatspothero commented:
@findepi the secrets run is green, see here: https://github.com/trinodb/trino/actions/runs/4344191372

findepi commented:
@grantatspothero thanks for reminding me we don't need a run with secrets for this PR
